hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/ReplicationServer.java

```diff
@@ -154,6 +155,31 @@ public int getPort() {
     return port;
   }
 
+  public void setPoolSize(int size) {
```

Comment:
There is logic in `nodeStateUpdated` that doubles the threads and queue size if the node is decommissioning or entering maintenance. I think that with this change, if the pool size had been doubled, this will not respect that.

Additionally, there is a function passed to ReplicationServer to do the update for that purpose. It is passed in from ReplicationSupervisorBuilder:

```
       executorThreadUpdater = threadCount -> {
          if (threadCount < tpe.getCorePoolSize()) {
            tpe.setCorePoolSize(threadCount);
            tpe.setMaximumPoolSize(threadCount);
          } else {
            tpe.setMaximumPoolSize(threadCount);
            tpe.setCorePoolSize(threadCount);
          }
        };
```
I think that after the config object is updated, we need to invoke the logic in `nodeStateUpdated` as that may cause the threads to be doubled in the node is decommissioning or entering maintenance. Maybe we need a force flag on that method so it always run even if the newState == oldState, or we need to extract the logic for whether to scale it up or not out into another method this reconfig could use.

Reply:
@guohao-rosicky can you please check the logic @sodonnel mentioned?

Reply:
There is one ReplicationServer, acting as the server to provide data to remote replication request,  and one ReplicationSupervisor, acting as the peer to start the replication request(either pull or push).  This patch adds the dynamic reconfiguration for ReplicationServer only, while the dynamic reconfiguration for ReplicationSupervisor is not covered.  @guohao-rosicky , I think you can add the support for ReplicationSupervisor in this patch, or you open a new JIRA to do that, both will be fine.

Reply:
Created HDDS-10237.